app/tests/src/test_chat_api.py

```diff
@@ -24,100 +23,113 @@
+    "This is by design and cannot be changed. If you change your loop between tests, make sure you do not reuse any pools or connections."
+    https://github.com/MagicStack/asyncpg/issues/293#issuecomment-391157754
+    Running a single test passes, but running all tests in this file fails.
+    This only affects tests, where the router functions are called by the FastAPI's TestClient, as opposed to sending HTTP requests.
+
+    TLDR: chat_api uses chainlit_data_layer, which uses asyncpg, which is tied to the event loop.
+    Since tests can run in different event loops, we need to reset the chainlit_data_layer between tests.
+
+    Alternative 1: to create a new FastAPI router for each test but router is referenced in many places
+        (including API endpoints setup) so some references may point to the original router.
+    Alternative 2: use one event loop for all tests -- https://github.com/pytest-dev/pytest-asyncio/issues/924#issuecomment-2328433273.
+    Alternative 3: Create a asyncio.new_event_loop(), asyncio.set_event_loop(new_loop),
+        new_loop.run_until_complete(_cause_asyncpg_create_pool()), and new_loop.run_until_complete(_my_test())
+    """
+    cl_data._data_layer = None
+    cl_data._data_layer_initialized = False
+
+
+@pytest.mark.asyncio
+async def test_api_engines(async_client):
+    response = await async_client.get("/api/engines?user_id=TestUser")
     assert response.status_code == 200
     assert response.json() == ["imagine-la"]
 
 
 @pytest.mark.asyncio
 async def test_api_engines__dbsession_contextvar(async_client, monkeypatch):
     event = threading.Event()
-    db_sessions = set()
+    db_sessions = []
+    orig_init_chat_session = chat_api._init_chat_session
 
-    async def waiting_get_or_create_user(_self, identifier, _metadata):
+    async def wait_for_all_requests(_self, *_args):
         db_session = chat_api.dbsession.get()
-        db_sessions.add(db_session)
+        db_sessions.append(db_session)
         # Run the blocking wait in a separate thread to avoid blocking the event loop
         await asyncio.to_thread(event.wait)
+        # Call original function
+        return await orig_init_chat_session(_self, *_args)
 
-        mock_user = MagicMock()
-        mock_user.id = f"litai_uuid_for_{identifier}"
-        return mock_user
-
-    monkeypatch.setattr(MockLiteralAIApi, "get_or_create_user", waiting_get_or_create_user)
+    monkeypatch.setattr(chat_api, "_init_chat_session", wait_for_all_requests)
 
     async def checker_coroutine():
-        # Allow the event loop to run other tasks
-        await asyncio.sleep(0)
-        # chat_api.dbsession.get() should be distinct for each thread
-        assert len(db_sessions) == 2
+        while len(db_sessions) < 2:
+            print(
+                "Checker coroutine waiting for other tasks to add to dbsessions", len(db_sessions)
+            )
+            # Allow the event loop to run other tasks
+            await asyncio.sleep(0.1)
```

Comment:
Previous code was wrong. Need a while loop to check and wait before checking any assertion. Also, the assertion is moved to line 146.